app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -311,7 +313,7 @@ class PipelineViz extends React.Component {
   }
 
   render() {
-    const stageContainers = this.stageNames.map((stageName, i) => {
+    let stageContainers = this.stageNames.map((stageName, i) => {
```

Comment:
This is getting harder read. Can you split into
* a function that render container
* a function that render stage arrow
* render that then loops through the stages and calls `renderStageArrow` for the first stage and `renderContainer` for all the others
?